### PR TITLE
feat(go binding): add beamsize/entropythold/maxcontext to context interface

### PR DESCRIPTION
**Pull request by hsinhoyeh**
fixes: #2349

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]

**Comment by hsinatfootprintai** [comment]
@ggerganov I think the build failed for go building has been fixed. please check it.

**Comment by hsinhoyeh** [comment]
related issue: #2330 #2339

**Comment by hsinatfootprintai** [comment]
@ggerganov seems it didn't re-test automatically, require manually trigger

**Comment by ggerganov** [comment]
Just triggered it again - let's see if it's green

**Comment by hsinhoyeh** [comment]
@ggerganov it is green now. it seems that we need a static library `libwhisper.a` as well as a dynamic library `libggml.so`.
is it intentional? making a clean path for lamma.cpp I guess?

**Comment by ggerganov** [comment]
> it seems that we need a static library libwhisper.a as well as a dynamic library libggml.so.

In what sense we need both?

**Comment by hsinhoyeh** [comment]
> > it seems that we need a static library libwhisper.a as well as a dynamic library libggml.so.
> 
> In what sense we need both?

okey, so it is clear that this is NOT intentional...

**Comment by hsinhoyeh** [comment]
> > it seems that we need a static library libwhisper.a as well as a dynamic library libggml.so.
> 
> In what sense we need both?

@ggerganov this CL makes much more sense now.

File: Makefile [hunk 1]
```diff
@@ -967,11 +967,14 @@ src/whisper.o: \
 
 $(LIB_WHISPER): \
 	$(OBJ_WHISPER) \
+	$(OBJ_GGML) \
```
Comment (ggerganov) [inline Makefile:970]: I don't think it is necessary to include the `ggml` object files in the `whisper` dynamic lib. We want to link `ggml` dynamically

Comment (hsinatfootprintai) [reply Makefile]: ya, removed.

**Comment by hsinhoyeh** [comment]
@ggerganov This issue may require your attention, as similar issues with Go bindings keep recurring.

File: Makefile [hunk 2]
```diff
@@ -971,7 +971,9 @@ $(LIB_WHISPER): \
 	$(CXX) $(CXXFLAGS) -shared -fPIC -o $@ $^ $(LDFLAGS)
 
 $(LIB_WHISPER_S): \
-	$(OBJ_WHISPER)
+	$(OBJ_WHISPER) \
+	$(OBJ_GGML) \
+	$(LIB_GGML)
```
Comment (ggerganov) [inline Makefile:976]: Is this change necessary?

Comment (hsinhoyeh) [reply Makefile:976]: ya, otherwise it can't find `ggml_xxx` library

error message like:

```
/usr/bin/ld: whisper.cpp:(.text+0x2a82b): undefined reference to `ggml_time_us'
/usr/bin/ld: whisper.cpp:(.text+0x2a881): undefined reference to `ggml_time_us'
/usr/bin/ld: whisper.cpp:(.text+0x2b191): undefined reference to `ggml_time_us'
/usr/bin/ld: whisper.cpp:(.text+0x2b2e1): undefined reference to `ggml_time_us'
/usr/bin/ld: /home/ubuntu/xxxx/libwhisper.a(whisper.o):whisper.cpp:(.text+0x2b5fe): more undefined references to `ggml_time_us' follow

```
